### PR TITLE
Ens changes + module changes

### PR DESCRIPTION
**Pull request by rahul7668gupta**
unit tests after extensive coverage
Screenshot 2023-02-08 at 2 29 01 AM

### CODE REVIEW

File: airstack-modules/modules/airstack/domain-name/domain-name.ts [hunk 1]
```diff
@@ -180,178 +144,153 @@ export namespace domain {
...
-    if (fromOldRegistry && node != ROOT_NODE && isMigratedMapping.isMigrated == true) {
-      // this domain was migrated from the old registry, so we don't need to hanlde old registry event now
-      return;
-    }
+    let domain = getOrCreateAirDomain(new Domain(domainId, chainId, airBlock, tokenAddress));
     // get previous resolver
     let previousResolverId = domain.resolver;
     // create new resolver
-    let resolverEntity = getOrCreateAirResolver(domain, chainId, resolver);
+    let resolverEntity = getOrCreateAirResolver(domain, chainId, airBlock, resolver);
```
Comment (0xsarvesh) [inline airstack-modules/modules/airstack/domain-name/domain-name.ts:170]: are you using airBlock is resolver entity?

File: airstack-modules/modules/airstack/nft-marketplace/nft.ts [hunk 2]
```diff
@@ -42,9 +42,9 @@ export namespace nft {
             let paymentToken = getOrCreateAirToken(chainID, NftSales[i].paymentToken.toHexString());
 
             // Account
-            let buyerAccount = handleAccountCreation(chainID, NftSales[i].buyer.toHexString(), block.id);
-            let sellerAccount = handleAccountCreation(chainID, NftSales[i].seller.toHexString(), block.id);
-            let feeAccount = handleAccountCreation(chainID, NftSales[i].protocolFeesBeneficiary.toHexString(), block.id);
+            let buyerAccount = getOrCreateAirAccount(chainID, NftSales[i].buyer.toHexString(), block);
```
Comment (0xsarvesh) [inline airstack-modules/modules/airstack/nft-marketplace/nft.ts:45]: Where are you saving these accounts?

File: ens/subgraph.yaml [hunk 3]
```diff
@@ -119,119 +119,10 @@ dataSources:
         - event: NameRenewed(indexed uint256,uint256)
           handler: handleNameRenewed
   - kind: ethereum/contract
-    name: Resolver2
+    name: Resolver
     network: mainnet
     source:
       abi: Resolver
-      address: '0x4976fb03C32e5B8cfe2b6cCB31c09Ba78EBaBa41'
```
Comment (0xsarvesh) [inline ens/subgraph.yaml:126]: Why is this removed?

**Review by 0xsarvesh — CHANGES_REQUESTED** [review]

File: airstack-modules/modules/airstack/domain-name/domain-name.ts (hunk 1 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/domain-name/domain-name.ts:170]: air block is being used to create air account, and not being saved inside resolver

File: airstack-modules/modules/airstack/nft-marketplace/nft.ts (hunk 2 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/nft-marketplace/nft.ts:45]: fixing this, by saving inside the air account function

File: ens/subgraph.yaml (hunk 3 above)
Comment (rahul7668gupta) [reply ens/subgraph.yaml:126]: this is removed coz any domain can have any resolver and we cannot limit to and address, that's why we are listening to all the events of such signature.
that's why we were also doing `domain.resolver = resolver.id` in resolver events.

File: airstack-modules/modules/airstack/nft-marketplace/nft.ts (hunk 2 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/nft-marketplace/nft.ts:45]: done, rarible subgraph integration, testing and deployment in #43

File: airstack-modules/bin/index.ts [hunk 4]
```diff
@@ -69,211 +69,18 @@ for (let index = START_INDEX; index < args.length; index++) {
...
-          NftSales: Sale[],
-          protocolType: string,
-          protocolActionType: string,
-          timestamp: BigInt,
-          blockHeight: BigInt,
-          blockHash: string
-        )
-
-        For documentation and examples please check: https://github.com/Airstack-xyz/Subgraphs
+Integration for DOMAIN_NAME vertical done. For documentation and examples please check below link:
```
Comment (deepesh-kn) [inline airstack-modules/bin/index.ts:75]: It should be NFT marketplace.

File: airstack-modules/bin/index.ts [hunk 5]
```diff
@@ -69,211 +69,18 @@ for (let index = START_INDEX; index < args.length; index++) {
...
+https://github.com/Airstack-xyz/Subgraphs/blob/ens-airstack-module-refactor/airstack-modules/modules/airstack/nft-marketplace/Readme.md
```
Comment (deepesh-kn) [inline airstack-modules/bin/index.ts:76]: The URL should be pointing to main branch.

File: airstack-modules/modules/airstack/common/index.ts [hunk 6]
```diff
@@ -17,9 +18,9 @@ export const BIG_INT_ZERO = BigInt.fromI32(0);
 
 export const SUBGRAPH_SCHEMA_VERSION = "1.0.0";
 
-export const SUBGRAPH_NAME = "AIRSTACK-SUBGRAPH";
-export const SUBGRAPH_VERSION = "AIRSTACK-SUBGRAPH";
-export const SUBGRAPH_SLUG = "AIRSTACK-SUBGRAPH";
+export const SUBGRAPH_NAME = "ens";
+export const SUBGRAPH_VERSION = "v1";
+export const SUBGRAPH_SLUG = "ens-v1";
```
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/common/index.ts:23]: In common file, why do we have `ens` specific values? How this will be used?

File: airstack-modules/modules/airstack/common/index.ts [hunk 7]
```diff
@@ -44,12 +45,42 @@ AIR_NETWORK_MAP.set("osmosis", "OSMOSIS");
...
+AIR_CHAIN_ID_MAP.set("matic", "137");
+AIR_CHAIN_ID_MAP.set("xdai", "100");
+
 export function processNetwork(network: string): string {
   const value = AIR_NETWORK_MAP.get(network);
   const result: string = value !== null ? value : "unknown";
   return result;
 }
 
+export function processChainId(): string {
```
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/common/index.ts:77]: This is not doing any process, its just returning the network value.
Should this be rename to `getChainId()` ?

File: airstack-modules/modules/airstack/common/index.ts (hunk 6 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/common/index.ts:23]: done

File: airstack-modules/modules/airstack/common/index.ts (hunk 7 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/common/index.ts:77]: then we shuold also change the name of processNetwork function

File: airstack-modules/bin/index.ts (hunk 4 above)
Comment (rahul7668gupta) [reply airstack-modules/bin/index.ts]: done

File: airstack-modules/bin/index.ts (hunk 5 above)
Comment (rahul7668gupta) [reply airstack-modules/bin/index.ts]: done

File: airstack-modules/modules/airstack/domain-name/domain-name.ts [hunk 8]
```diff
@@ -643,6 +556,29 @@ export namespace domain {
   }
 
   // end of track functions and start of get or create and helper functions
+  /**
+   * @dev this function creates an air extra entity
+   * @param name air extra name
+   * @param value air extra value
+   * @param domainId air domain id
```
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/domain-name/domain-name.ts:563]: You can rename it as id and move the function in to common.

File: airstack-modules/modules/airstack/domain-name/domain-name.ts [hunk 9]
```diff
@@ -751,32 +691,34 @@ export namespace domain {
...
    * @returns entity id in string format
    */
-  function createEntityId(transactionHash: string, blockHeight: BigInt, logIndex: BigInt): string {
-    return transactionHash.concat("-").concat(blockHeight.toString()).concat('-').concat(logIndex.toString());
+  function createEntityId(transactionHash: string, blockHeight: BigInt, logOrCallIndex: BigInt): string {
+    return transactionHash.concat("-").concat(blockHeight.toString()).concat('-').concat(logOrCallIndex.toString());
   }
 
   /**
+   * @dev this function needs to be removed
```
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/domain-name/domain-name.ts:702]: Removed?

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
I feel that the code quality should be improved. 
We should re evaluate when we should save the entities.

File: airstack-modules/modules/airstack/domain-name/domain-name.ts [hunk 10]
```diff
@@ -19,99 +20,71 @@ import {
...
-    let isMigratedMapping = getOrCreateIsMigratedMapping(domainId, block.id, isMigrated);
-    if (fromOldRegistry && isMigratedMapping.isMigrated == true) {
-      // this domain was migrated from the old registry, so we don't need to hanlde old registry event now
-      return;
-    }
+    let chainId = processChainId();
+    let airBlock = getOrCreateAirBlock(chainId, block.number, block.hash.toHexString(), block.timestamp);
+    let domain = getOrCreateAirDomain(new Domain(domainId, chainId, airBlock, tokenAddress));
     let previousOwnerId = domain.owner;
     let parent = getOrCreateAirDomain(new Domain(
```
[the remark this answers is not in the document]
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/domain-name/domain-name.ts:64]: this getOrCreate function saves the entity if not there, and parent domain is of type AirDomain, so, when a parent domain owner changed txn is recorded, it gets created.

File: airstack-modules/modules/airstack/domain-name/domain-name.ts (hunk 8 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/domain-name/domain-name.ts:563]: cannot be done as all air extras dont have a domain id linked to them, social module has user id, so the air extra entity schema differs

File: airstack-modules/modules/airstack/domain-name/domain-name.ts (hunk 9 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/domain-name/domain-name.ts:702]: removed as this function is no longer being used

Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/domain-name/domain-name.ts:702]: removed makeSubnode function also

File: airstack-modules/modules/airstack/domain-name/domain-name.ts [hunk 11]
```diff
@@ -19,99 +20,71 @@ import {
...
+    ownerAccount.save();
+    domain.owner = ownerAccount.id;
     domain.parent = parent.id;
-    domain.labelHash = label;
+    domain.labelHash = labelHash;
     domain.tokenId = tokenId;
-    domain.lastBlock = block.id;
-    recurseSubdomainCountDecrement(domain, chainId, block, tokenAddress);
+    domain.lastBlock = airBlock.id;
+    recurseSubdomainCountDecrement(domain, chainId, airBlock, tokenAddress);
```
[the remark this answers is not in the document]
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/domain-name/domain-name.ts:82]: Found explanation in this issue/pr from the ens subgraph GitHub repo - https://github.com/ensdomains/ens-subgraph/pull/44
this is because when you change the resolver for a domain, your subdomains are no longer linked to the same resolver.
this was also discussed with @0xsarvesh in https://github.com/Airstack-xyz/Subgraphs/pull/34#discussion_r1096798600

File: airstack-modules/modules/airstack/common/index.ts (hunk 7 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/common/index.ts:77]: done

File: airstack-modules/modules/airstack/domain-name/domain-name.ts [hunk 12]
```diff
@@ -920,13 +872,13 @@ export namespace domain {
...
     blockHeight: BigInt,
-    logIndex: BigInt,
+    logOrCallIndex: BigInt,
     oldTTL: BigInt | null,
     newTTL: BigInt,
     block: AirBlock,
     domain: AirDomain,
   ): AirDomainNewTTLTransaction {
-    let id = createEntityId(transactionHash, blockHeight, logIndex);
+    let id = createEntityId(transactionHash, blockHeight, logOrCallIndex);
```
[the remark this answers is not in the document]
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/domain-name/domain-name.ts:859]: i have updated all the .save() calls outside the create funcs.

**Comment by rahul7668gupta** [comment]
@deeepesh-kn , requesting re-review for this pr.

File: airstack-modules/modules/airstack/domain-name/domain-name.ts [hunk 13]
```diff
@@ -123,52 +97,43 @@ export namespace domain {
...
       entity = new AirDomainTransferTransaction(id);
       entity.from = previousOwnerId;
-      entity.to = getOrCreateAirAccount(chainId, newOwnerAddress).id;
-      let airBlock = getOrCreateAirBlock(
-        chainId,
-        blockHeight,
-        blockHash,
-        blockTimestamp,
-      );
+      let newOwnerAccount = getOrCreateAirAccount(chainId, newOwnerAddress, airBlock);
```
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/domain-name/domain-name.ts:134]: You already have this enitity on line # 129

File: airstack-modules/modules/airstack/domain-name/domain-name.ts [hunk 14]
```diff
@@ -123,52 +97,43 @@ export namespace domain {
...
+      let previousOwnerAccount = getOrCreateAirAccount(chainId, ZERO_ADDRESS, airBlock);
+      previousOwnerAccount.save();
+      previousOwnerId = previousOwnerAccount.id;
     }
-    let id = createEntityId(transactionHash, blockHeight, logIndex);
+    // update domain owner here
+    let newOwnerAccount = getOrCreateAirAccount(chainId, newOwnerAddress, airBlock);
+    newOwnerAccount.save();
+    domain.owner = newOwnerAccount.id;
+    domain.save();
```
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/domain-name/domain-name.ts:128]: `lastBlock` should be updated?
it should be `lastUpdatedBlock`

File: ens/modules/airstack/domain-name/domain-name.ts [hunk 15]
```diff
@@ -461,26 +386,27 @@ export namespace domain {
...
+      airBlock,
       tokenAddress,
     ));
 
     // tracking registration cost in domain entity  - renewal cost is not being tracked yet
     if (fromRegistrationEvent) {
       domain.registrationCost = cost;
       domain.paymentToken = getOrCreateAirToken(chainId, paymentToken).id;
-      domain.lastBlock = block.id;
+      domain.lastBlock = airBlock.id;
```
Comment (deepesh-kn) [inline ens/modules/airstack/domain-name/domain-name.ts:401]: This should be out side if-else

File: airstack-modules/modules/airstack/domain-name/domain-name.ts [hunk 16]
```diff
@@ -1102,12 +1042,12 @@ export namespace domain {
    */
   function recurseSubdomainCountDecrement(domain: AirDomain, chainId: string, block: AirBlock, tokenAddress: string): string | null {
```
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/domain-name/domain-name.ts:1043]: Please verify if the return values are needed? 
I dont see its being used.

If `domain.id` is needed, can you check which `id` should should be returned? Its a recursive call.

File: airstack-modules/modules/airstack/domain-name/domain-name.ts [hunk 17]
```diff
@@ -180,279 +148,260 @@ export namespace domain {
...
+    if (domain.extras == null) {
+      extrasArray.push(extra.id);
+    } else {
+      extrasArray = domain.extras!;
+      extrasArray.push(extra.id);
     }
-    // update domain ttl
-    domain.ttl = newTTL;
-    domain.lastBlock = block.id;
+    domain.extras = extrasArray;
```
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/domain-name/domain-name.ts:245]: Evaluate this very carefully, the `extras` could have more that just TTL, we do not want to loose it with new array assignment.
In this case I think its fine, but still I think we should have the complete code that takes care of such scenario.
May be have a util function that can handle this.

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
@rahul7668gupta , over all looks good.
I have one comment, please check.

Also please verify that with the changes done to not save the entities in the `getOrCreate*` function, its handled in the other verticals, and it will not cause any problem there.

File: airstack-modules/modules/airstack/domain-name/domain-name.ts (hunk 16 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/domain-name/domain-name.ts]: removed return statements.

File: airstack-modules/modules/airstack/domain-name/domain-name.ts (hunk 17 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/domain-name/domain-name.ts:247]: it could be the case in socials subgraphs, and the module there is handling extras dynamically. Here, in domain-name, the function which uses extras is `trackDomainNewTTLTransaction` and only records ttl data. This will work and is covered in unit tests. Have also handled multiple ttl events for same extras entity in the test case now.

File: airstack-modules/modules/airstack/domain-name/domain-name.ts (hunk 16 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/domain-name/domain-name.ts]: done

File: airstack-modules/modules/airstack/domain-name/domain-name.ts (hunk 17 above)
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/domain-name/domain-name.ts:247]: done

**Review by deepesh-kn — APPROVED** [review]
